rp2040-hal/src/pio.rs

```diff
@@ -118,6 +118,18 @@ impl<P: PIOExt> PIO<P> {
         &self.interrupts
     }
 
+    /// Enable interrupts raised by state machines.
+    ///
+    /// The PIO peripheral has 4 outside visible interrupts that can be raised by the state machines. Note that this
+    /// does not correspond with the state machine index; any state machine can raise any one of the four interrupts.
```

Comment:
This description confuses me. (And, to be honest, the datasheet confuses me, too.)
Let me explain what I _think_ how the PIO interrupts work:

There are 8 interrupt flags which can be accessed from all state machines. That are the 8 possible values which can be set or queried with the `IRQ` or `WAIT` instructions, and that are outside visible in the PIO IRQ register.

4 of them can be routed to two externally visible interrupts, which you call `PioIRQ::Irq0` and `PioIRQ::Irq1`.
Those four are, a little bit confusingly, called `sm0`—`sm3` in the pio interrupt registers (`IRQ0_INTE` etc.), but do _not_ correspond to the 4 state machines.

If that's correct, I would not call it "4 outside visible interrupts", I only see two interrupts from the outside. But they can be connected to 4 of the 8 interrupt flags visible from inside the PIO.

Reply:
Your understanding matches my confused understanding of the datasheet.